### PR TITLE
Move `FlowSummaryImpl.qll` to `dataflow` pack

### PR DESCRIPTION
**Pull request by hvitved**
This PR turns `FlowSummaryImpl.qll` (and `AccessPathSyntax.qll`) into a proper parameterized module, and moves it to the shared `dataflow` pack.

Previously, the exposed `SummarizedCallable` class would look like
```ql
abstract class SummarizedCallable extends SummarizedCallableBase {
  predicate propagatesFlow(
    SummaryComponentStack input, SummaryComponentStack output, boolean preservesValue
  ) {
    none()
  }

  predicate propagatesFlowExt(
    string input, string output, boolean preservesValue
  ) {
    none()
  }
}
```

but after this PR it is simply

```ql
abstract class SummarizedCallable extends SummarizedCallableBase {
  abstract predicate propagatesFlow(
    string input, string output, boolean preservesValue
  );
}
```

That is, we now consider `SummaryComponentStack` an internal implementation detail, which is not exposed to the user (the user-facing aliases have been deprecated).

Commit-by-commit review is encouraged, and for each of the `<LANG>: Use FlowSummaryImpl from dataflow pack` commits, I recommend viewing the `FlowSummaryImpl.qll` in full, as the diff is not really meaningful (previously, it was the synced copy of share library, now it contains the inputs to the shared library).

### CODE REVIEW

**Review by erik-krogh — APPROVED** [review]
JS 👍

**Review by MathiasVP — APPROVED** [review]
Swift 👍

File: python/ql/lib/semmle/python/frameworks/data/internal/ApiGraphModels.qll [hunk 1]
```diff
@@ -70,8 +70,8 @@ private module API = Specific::API;
 
 private module DataFlow = Specific::DataFlow;
 
-private import Specific::AccessPathSyntax
 private import ApiGraphModelsExtensions as Extensions
+import codeql.dataflow.internal.AccessPathSyntax
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/data/internal/ApiGraphModels.qll:74]: Can this import be made private?

```suggestion
private import codeql.dataflow.internal.AccessPathSyntax
```

File: python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll [hunk 2]
```diff
@@ -1559,12 +1558,15 @@ private class SummaryReturnNode extends FlowSummaryNode, ReturnNode {
 }
 
 private class SummaryArgumentNode extends FlowSummaryNode, ArgumentNode {
+  private SummaryCall call_;
+  private ArgumentPosition pos_;
+
   SummaryArgumentNode() {
-    FlowSummaryImpl::Private::summaryArgumentNode(_, this.getSummaryNode(), _)
+    FlowSummaryImpl::Private::summaryArgumentNode(call_.getReceiver(), this.getSummaryNode(), _)
```
[the remark this answers is not in the document]
Comment (hvitved) [reply python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll:1565]: Yes, good catch. Is it the case that Python does not yet have flow summary models with callbacks? Otherwise I would have expected this CP to show up in DCA.

File: python/ql/lib/semmle/python/frameworks/data/internal/ApiGraphModels.qll (hunk 1 above)
Comment (hvitved) [reply python/ql/lib/semmle/python/frameworks/data/internal/ApiGraphModels.qll]: Yes, but it requires an extra import in `ApiGraphModelsSpecific.qll`, but that's ok.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/ExternalFlow.qll [hunk 3]
```diff
@@ -485,3 +504,63 @@ string asPartialNeutralModel(UnboundCallable c) {
...
+) {
+  exists(
+    string namespace, string type, boolean subtypes, string name, string signature, string ext
+  |
+    summaryModel(namespace, type, subtypes, name, signature, ext, input, output, kind, provenance) and
+    c = interpretElement(namespace, type, subtypes, name, signature, ext)
+  )
+}
+
+private class SummarizedCallableAdapter extends SummarizedCallable {
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/ExternalFlow.qll:519]: Maybe a comment that this is where the MaD rows are being used to create SummarizedCallable's (the same for neutrals as well).

**Review by michaelnebel — APPROVED** [review]
C#: Looks good to me!
Thank you for doing this @hvitved !

File: python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll]: We don't have too many flow summary models yet, so it could certainly explain it :+1:

**Review by RasmusWL — DISMISSED** [review]
Python :+1: (thanks for the updates)

**Review by yoff — COMMENTED** [review]
I can see this has already been approved by Python. I can also see that I have a pending comment, but I do not remember what it is, so I will just submit the partial review now, to see if it is still relevant.

File: shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll [hunk 4]
```diff
@@ -0,0 +1,1912 @@
...
+        mid.drop(mid.length() - 2) =
+          SummaryComponentStack::push(TParameterSummaryComponent(_),
+            SummaryComponentStack::singleton(TArgumentSummaryComponent(_))) and
+        preservesValue = preservesValue1.booleanAnd(preservesValue2)
+      )
+    }
+
+    /**
+     * Holds if `c` has a flow summary from `input` to `arg`, where `arg`
+     * writes to (contents of) arguments at position `pos`, and `c` has a
```
[the remark this answers is not in the document]
Comment (yoff) [reply shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll:655]: Aha! Not particularly critical...

**Comment by aschackmull** [comment]
I think I might see a bug related to the MaD provenance column. Could you verify that the presence of both a manual and a generated model for the same callable yields only the manual summary? I believe this used to work, but I think it might be broken now.

File: shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll [hunk 5]
```diff
@@ -0,0 +1,1912 @@
...
+      signature module SourceSinkInterpretationInputSig<LocationSig Location> {
+        class Element {
+          string toString();
+
+          Location getLocation();
+        }
+
+        /**
+         * Holds if an external source specification exists for `n` with output specification
+         * `output`, kind `kind`, and provenance `provenance`.
```
Comment (aschackmull) [inline shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll:1495]: There's no provenance column.
```suggestion
         * `output`, and kind `kind`.
```

File: shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll [hunk 6]
```diff
@@ -0,0 +1,1912 @@
...
+         */
+        predicate sourceElement(Element n, string output, string kind);
+
+        /**
+         * Holds if an external sink specification exists for `n` with input specification
+         * `input`, kind `kind` and provenance `provenance`.
```
Comment (aschackmull) [inline shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll:1501]:
```suggestion
         * `input`, and kind `kind`.
```

**Comment by owen-mc** [comment]
@aschackmull Do we have a test for what happens when we have a manual model and a generated model?

**Comment by hvitved** [comment]
> Do we have a test for what happens when we have a manual model and a generated model?

There is at least https://github.com/github/codeql/blob/main/java/ql/src/Metrics/Summaries/GeneratedVsManualCoverage.ql.

File: shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll [hunk 7]
```diff
@@ -0,0 +1,1912 @@
...
+     *     none()
+     *   }
+     *
+     *   override predicate hasProvenance(Provenance provenance) {
+     *     C.super.hasProvenance(provenance)
+     *   }
+     * }
+     * ```
+     */
+    abstract class SummarizedCallableImpl extends SummarizedCallableBaseFinal {
```
Comment (aschackmull) [inline shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll:570]: What does this class achieve? It's instantiation in `SummarizedCallableExternal` seems somewhat nonsensical since the `propagatesFlow` columns and `provenance` columns are formed by cartesian product. Aka, I don't think the generated vs manual logic in `SummarizedCallableExternal` does anything? That's all been moved to the language-specific side, right?

Comment (aschackmull) [reply shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll:570]: Ah, right, I see that it handles the translation between input/output-specs of type string vs. `SummaryComponentStack`, so I'll rephrase this as a simplification comment in `SummarizedCallableExternal`.

File: shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll [hunk 8]
```diff
@@ -0,0 +1,1912 @@
...
+        head = interpretComponent(spec.getToken(n - 1))
+      }
+
+      private class MkStack extends RequiredSummaryComponentStack {
+        override predicate required(SummaryComponent head, SummaryComponentStack tail) {
+          interpretSpec(_, _, head, tail)
+        }
+      }
+
+      private class SummarizedCallableExternal extends SummarizedCallableImpl instanceof SummarizedCallable
```
Comment (aschackmull) [inline shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll:1426]: I think this can be simplified. If you consider inlining `summaryElement` then several pieces of this class falls apart. I.e. `relevantSummaryElement` is exactly equal to `super.propagatesFlow`, so that can be deleted and replaced, which also means that `relevantSummaryElementGenerated` can be deleted. The charpred can also be removed, I think, since that's basically represented by the `instanceof SummarizedCallable`.

File: shared/dataflow/codeql/dataflow/internal/AccessPathSyntax.qll [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+/**
+ * Parses a lower-bounded interval `n..` and gets the lower bound.
+ */
+bindingset[arg]
+int parseLowerBound(string arg) { result = arg.regexpCapture("(-?\\d+)\\.\\.", 1).toInt() }
+
+/**
+ * An access part token such as `Argument[1]` or `ReturnValue`.
```
Comment (aschackmull) [inline shared/dataflow/codeql/dataflow/internal/AccessPathSyntax.qll:40]:
```suggestion
 * An access path token such as `Argument[1]` or `ReturnValue`.
```

**Comment by michaelnebel** [comment]
> @aschackmull Do we have a test for what happens when we have a manual model and a generated model?

I don't there as such exist one for Java (but there exist one for C#).
One can use the [summary/1](https://github.com/github/codeql/blob/1653433f39f1632b9853062297c8fae5c4fea3d9/java/ql/lib/semmle/code/java/dataflow/internal/FlowSummaryImpl.qll#L1364) predicate for creating such a test.

File: shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll (hunk 7 above)
Comment (hvitved) [reply shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll:570]: Indeed.

File: shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll (hunk 8 above)
Comment (hvitved) [reply shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll:1426]: Indeed, this is leftover from the old implementation, thanks for spotting.

**Review by aschackmull — APPROVED** [review]
LGTM! Thanks for tackling this!

**Review by aibaars — APPROVED** [review]
